py/redrock/utils.py

```diff
@@ -216,26 +216,82 @@ def distribute_work(nproc, ids, weights=None, capacities=None):
     return dist
 
 
-def transmission_Lyman(zObj,lObs):
+def transmission_Lyman(zObj,lObs, use_gpu=False):
     """Calculate the transmitted flux fraction from the Lyman series
     This returns the transmitted flux fraction:
     1 -> everything is transmitted (medium is transparent)
     0 -> nothing is transmitted (medium is opaque)
+
+    This method will handle 3 options:
+    1 -> For GPU mode, zObj is an array of all z for a template and the return
+    value will be a cupy array (nz x nlambda)
+    2-> In CPU mode, it can auto-detect if zObj is a numpy array and if so,
+    again, all z will be processed as a vector and the return value
+    will be a numpy array (nz x nlambda)
+    3-> For legacy, it is still supported to pass zObj as a float and
+    in this case, the return value will be a 1-d numpy array (nlambda).
+
     Args:
-        zObj (float): Redshift of object
+        zObj (float or array of float): Redshift(s) of object
         lObs (array of float): wavelength grid
     Returns:
         array of float: transmitted flux fraction
-    """
 
-    lRF = lObs/(1.+zObj)
-    T   = np.ones(lObs.size)
+    """
 
     Lyman_series = constants.Lyman_series
+    if (use_gpu):
+        #Use GPU helper function
+        return transmission_Lyman_gpu(zObj, lObs)
+    elif (type(zObj) == np.ndarray):
+        #This is an array of float
+        if (lObs.min()/(1+zObj.max()) > Lyman_series['Lya']['line']):
+            #Return None if wavelength range doesn't overlap with Lyman series
+            #No need to perform any calculations in this case
+            return None
+        lObs = np.tile(lObs, (zObj.size, 1))
+        lRF = lObs/(1.+np.array(zObj)[:,None])
+        T = np.ones(lRF.shape)
+    else:
+        #zObj is a float
+        lRF = lObs/(1.+zObj)
+        T   = np.ones(lObs.size)
+
     for l in list(Lyman_series.keys()):
         w      = lRF<Lyman_series[l]['line']
         zpix   = lObs[w]/Lyman_series[l]['line']-1.
         tauEff = Lyman_series[l]['A']*(1.+zpix)**Lyman_series[l]['B']
         T[w]  *= np.exp(-tauEff)
 
     return T
+
+def transmission_Lyman_gpu(all_zObj,lObs):
+    """Calculate the transmitted flux fraction from the Lyman series
+    This returns the transmitted flux fraction:
+    1 -> everything is transmitted (medium is transparent)
+    0 -> nothing is transmitted (medium is opaque)
+    Args:
+        all_zObj (array of float): Redshift of all objects
+        lObs (array of float): wavelength grid
+    Returns:
+        cp.array of float (nz x nlambda): transmitted flux fraction
+    """
+    import cupy as cp
```

Comment:
There is a lot of replicated code in `transmission_Lyman_gpu` compared to `transmission_Lyman`, differing only by `np` vs. `cp`.  Let's use the API similarity between those two libraries to reduce this to a single function that automatically does the right thing with either np or cp arrays.  Similarly the scalar vs. array `zObj` can be handled with `np.atleast_1d` to promote it to an array, do the calculation, and then special case whether the return value should be sliced back to a 1D array or left as a 2D array.

Reply:
The updated `transmission_Lyman` resolves these concerns, and is a good model for future maintainability of CPU vs. GPU code paths — it's the same code for both with no duplication of core algorithm statements, and very minimal if/else handling of data types for the two cases.